### PR TITLE
fix: boolean type into SQL 'in' operator

### LINKED ISSUE #16081: [cross-filtering] treemap not emitting boolean column correctly
Emitting boolean filter from treemap is wrong

### Expected results
```WHERE `bool_automated` IN 'true)```

### Actual results
```WHERE `bool_automated` IN ('true')```

#### Screenshots
https://user-images.githubusercontent.com/52086618/128303930-898b9f6c-7b34-4295-9e25-a70af8e11732.mp4

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

closes: https://github.com/apache/superset/issues/16081
Currently, there is an issue with filters that handle boolean values. If the database engine does not provide boolean value implicit conversion, will cause the query to fail.
Imagine SQL:
```
SELECT case
           when gender = 'boy' then True
           else False
       end AS boolean_gender,
       count(num) AS "COUNT(num)"
FROM birth_names
WHERE case
          when gender = 'boy' then True
          else False
      end IN ('false')
GROUP BY case
             when gender = 'boy' then True
             else False
         end
ORDER BY "COUNT(num)" DESC
LIMIT 10000;
```

the derived column(Calculated column in Superset) type is Boolean, and the operand is String('true')

```
case
    when gender = 'boy' then True
    else False
end IN ('false')
```

This PR introduced a conversion when the column type is boolean and operand type also is boolean.

#### Before

https://user-images.githubusercontent.com/2016594/128458491-8c8d4e31-efb7-4445-9e2f-7ac50d246b42.mov

### TESTING INSTRUCTIONS

1. select `birth_names` dataset
2. create a new Calculated Columns. 
name: boolean_gender
SQL expression : `case when gender = 'boy' then True else False end`
Data Type: Boolean
3. create a treemap v2. select `boolean_gender` into Group by control; select sum(num) into Metric; check `emit dashboard cross filters`; saved to new dashboard `test`
4. create a table. select `boolean_gender` into Group by control; select sum(num) into Metric; check `emit dashboard cross filters`; saved to dashboard `test`
5. goto dashboard `test`
6. select `true` or `false` in treemap v2, then the table can be filtered.
7. double-check view query in table viz, where clause like below:
```
WHERE case
          when gender = 'boy' then True
          else False
      end IN (false)
```

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/16081
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/datasource/DatasourceEditor.jsx [hunk 1]
```diff
@@ -112,7 +112,7 @@ const ColumnButtonWrapper = styled.div`
 const checkboxGenerator = (d, onChange) => (
   <CheckboxControl value={d} onChange={onChange} />
 );
-const DATA_TYPES = ['STRING', 'NUMERIC', 'DATETIME'];
+const DATA_TYPES = ['STRING', 'NUMERIC', 'DATETIME', 'BOOLEAN'];
```
Comment (zhaoyongjie) [inline superset-frontend/src/datasource/DatasourceEditor.jsx:115]: bycatch: added boolean type in the UI

File: superset/utils/core.py [hunk 2]
```diff
@@ -423,6 +423,35 @@ def cast_to_num(value: Optional[Union[float, int, str]]) -> Optional[Union[float
...
+    >>> cast_to_boolean(0)
+    False
+    >>> cast_to_boolean(0.5)
+    True
+    >>> cast_to_boolean('true')
+    True
+    >>> cast_to_boolean('false')
+    False
+    >>> cast_to_boolean('False')
+    False
```
Comment (villebro) [inline superset/utils/core.py:440]: Nit: I would also add this as it seems to be missing
```python
>>> cast_to_boolean('True')
True
```

**Review by villebro — APPROVED** [review]
LGTM with one non-blocking recommendation

**Comment by rosemarie-chiu** [comment]
🏷 2021.31

**Pull request merged**